guides/hack/08-generic-types-and-functions/20-reified-generics.md

```diff
@@ -0,0 +1,187 @@
+
+## Parameter and return type verification
+
+Prior to reified generics, if the type hint for a parameter or return value had generics, HHVM would only check whether the input is of the outermost type (whether `$c` is a `C` in the next example). With reified generics, HHVM will also check and enforce the reified type parameters.
+
+```Hack
+class C<reify T> {}
+
+function f(C<int> $c): void {}
+
+$c1 = new C<int>();
+f($c1); // success
+$c2 = new C<string>();
+f($c2); // parameter type hint violation
+```
+
+The reified type parameter is checked as well:
+
+```Hack
+class C<reify T> {}
+
+function f<reify T>(T $x): C<T> {
+  return new C<int>();
+}
+
+f<int>(1); // success
+f<int>("yep"); // parameter type hint violation
+f<string>("yep"); // return type hint violation
+```
+
+## Type testing and assertion with `is` & `as` expressions
+
+Suppose you have a `vec<mixed>` and you want to extract all types `T` from it. Prior to reified generics, you'd need to implement a new function for each type `T` but with reified generics you can do this in a generic way. Start by adding the keyword `reify` to the type parameter list.
+
+```Hack
+function filter<<<__Enforceable>> reify T>(vec<mixed> $list): vec<T> {
+  $ret = vec[];
+  foreach ($list as $elem) {
+    if ($x is T) {
+      $ret[] = $elem;
+    }
+  }
+  return $ret;
+}
+
+filter<int>(vec[1, "hi", true])
+// => vec[1]
+filter<string>(vec[1, "hi", true])
+// => vec["hi"]
+```
+
+Notice that the reified type parameter has the attribute `<<__Enforceable>>`. In order to use type testing and type assertion, the reified type parameter must be marked as `<<__Enforceable>>`, which means that we can fully enforce this type parameter, i.e. it does not contain any erased generics, not a function type, etc.
+
+```Hack
+class A {}
+class B<reify T> {}
+class C<reify Ta, Tb> {}
+
+int // enforceable
+A // enforceable
+B<int> // enforceable
+B<A> // enforceable
+C<int, string> // NOT enforceable as C's second generic is erased
+```
+
+## Creating an instance of a class with `new`
+
+Prior to reified generics, in order to create a new instance of a class without a constant class name, you'd need to pass it as `classname<T>` which is not type safe. In the runtime, classnames are strings.
+
+```Hack
+<<__ConsistentConstruct>>
+abstract class A {}
+
+class B extends A {}
+class C extends A {}
+
+function f<<<__Newable>> reify T as A>(): T {
+  return new T();
+}
+
+f<A>(); // not newable since it is abstract class
+f<B>(); // success
+f<C>(); // success
+```
+
+Notice that the reified type parameter has the attribute `<<__Newable>>`. In order for a type to be `<<__Newable>>`, the type must represent a class that's not abstract and has a consistent constructor or be a final class. Creating a new instance using the reified generics also carries across the generics given. For example,
```

Comment:
`__Newable` implies that `new T(/*__ConsistentConstructs params*/)` will be valid.
Does this also open up `is` and `as` checks? Or do you need a `__Newable __Enforceable T` if you with to use `new` and `is` / `as` ? `__Newable` already prevents HHVM from erasing the `T` at runtime.